s3-source-connector/src/main/java/io/aiven/kafka/connect/s3/source/utils/FileReader.java

```diff
@@ -95,13 +95,21 @@ private void fetchNextBatch() {
                         .stream()
                         .filter(objectSummary -> objectSummary.getSize() > 0)
                         .filter(objectSummary -> !failedObjectKeys.contains(objectSummary.getKey()))
+                        .filter(objectSummary -> assignObjectToTask(objectSummary.getKey()))
                         .collect(Collectors.toList());
 
                 continuationToken = objectListing.getNextContinuationToken();
                 isTruncated = objectListing.isTruncated();
 
                 currentBatch.forEach(objSummary -> LOGGER.debug("Objects to be processed {} ", objSummary.getKey()));
             }
+
+            private boolean assignObjectToTask(final String objectKey) {
+                final int maxTasks = Integer.parseInt(s3SourceConfig.originals().get("tasks.max").toString());
+                final int taskId = Integer.parseInt(s3SourceConfig.originals().get("task.id").toString()) % maxTasks;
+                final int taskAssignment = Math.floorMod(objectKey.hashCode(), maxTasks);
+                return taskAssignment == taskId;
+            }
```

Comment:
This seems to be a bit arbitrary, 
Here it evenly distributes the maximum number of tasks, but i think it would make sense to use a task partitioner. 

The task partitioner could look at the filename of the object e.g. {{topic}}-{{partition}}-{{start_offset}} pick out the partition and decide if it should be the one to process that partition. [docs](https://aiven.io/docs/products/kafka/kafka-connect/reference/s3-sink-additional-parameters)

Similarly for those that use other S3 Sinks such as confluents version the naming convention would be 
topics/<TOPIC_NAME>/partition=<PARTITION_NUMBER>/<TOPIC_NAME>+<PARTITIOIN_NUMBER>+<START_OFFSET>.<FILE_EXTENSION> [docs](https://aiven.io/docs/products/kafka/kafka-connect/reference/s3-sink-additional-parameters-confluent)

Partition again can be used and potentially more efficiently here because the partition is defined in the path meaning only on starting the task would this task partitioner need to be invoked.

That would allow us to execute the process for instance by partition, so each task would take a partition and start processing records, if the user has used key and value this should ensure event ordering. Currently if one task is slow or encounters longer latency because it is distributed in a different DC the events could get published back to kafka out of order.

Reply:
This particular way of distributing could be a separate task partitioner where the objects are not originally from another S3 Sink or the user does not care about ordering.

Reply:
Agree, it evenly distributes the amongst tasks. With this pr, I only want to fix the duplicate processing bug and with updated unit tests, I was making sure even if we have multiple tasks, there is no duplicate processing.

We have that task anyways for partitioner and depending on the configuration right strategy could be applied.

Reply:
This is a simple solution, until we have all those features implemented.

Reply:
And we should document this way of working until we have new ways implemented, so users are aware.

Reply:
Hi @muralibasani I was working on a tangential issue and found that this implementation had a second more difficult issue to resolve.

Currently if you have 5 threads and 3 partitions any task can process any file. The issue comes around when something like this happens,

task 1 process' file logs-0-0001.gz
task 3 process' file logs-0-0101.gz
task 1 process' file logs-0-0201.gz
task 3 tries to process' file logs-0-0301.gz
task 1 process' process' file logs-0-0401.gz
task 3 fails during offset 301 and falls over and restarts, but task 1 has updated the offsets saved in distributed mode to be set to 401 and then the events missed in  logs-0-0301.gz will never be processed.

Reply:
This will not be an issue, as positions are stored in backing store, and can be picked up by any task.
I will try to add tests with restarts on workers.